### DISCUSSION TITLE
TabItem.Header Style

### DISCUSSION THREAD

**Discussion by drei97** [AvaloniaUI/Avalonia#13951] [Q&A] [answered]
Hi, 

after upgrade from 11.0.5 to 11.0.6 my header of the selected TabItem is now completely blue. Fore- and background are of same color and not according former defaults.

![grafik](https://github.com/AvaloniaUI/Avalonia/assets/68755045/9f38b11f-27ea-4f46-b4ed-79f8cf2db4ce)

![grafik](https://github.com/AvaloniaUI/Avalonia/assets/68755045/f38b4cc9-3354-4ac9-a11c-51ee510c05e8)

Must be a bug?

3.97

**Comment by timunie** [comment]
well... without seeing any code hard to tell.

**Comment by drei97** [comment]
There is not much to show. I am using defaults only

MainWindow.xaml

```
        <Border Grid.Row="1"
                Classes="Content">
            <TabControl
                ItemsSource="{Binding Tabs}"
                SelectedItem="{Binding SelectedTab}">
            </TabControl>
        </Border>
```

the TabItems are created dynamically in the Mainwindows ViewModel:

```
   private void AngeboteListeÖffnen()
    {
        View!.Cursor = new Cursor(StandardCursorType.Wait);
        var item = Tabs.FirstOrDefault(xx => (string)xx.Header! == QuotesViewModel.PräfixPlural);

        if (item == null)
        {
            item = new TabItem { Header = QuotesViewModel.PräfixPlural };
            var content = new QuotesView();
            item.Content = content;
            Tabs.Add(item);
        }

        SelectedTab = item;
        View.Cursor = Cursor.Default;
    }
```

I am using no own styles for TabItems, since I am really no expert in that matter.

After downgrading back to 11.0.5 it is now working as expected.

![grafik](https://github.com/AvaloniaUI/Avalonia/assets/68755045/31c917a9-06bb-4390-affb-31141b30ce51)

something has changed in 11.0.6

**Reply by timunie** [reply]
You should not create a TabItem inside your ViewModel. It should always be done by DataTemplates or by override ToString

**Reply by drei97** [reply]
Why? It is working perfectly so far. The number of tabs is not constant and the tabs are representing docked windows which can be opened and closed by the user. How should this be done in XAML to handle a non constant number of tabs?
And this cannot be any reason for the change I see in my project.

In 11.0.5 the selected tab is underlined by a thin blue line. In 11.0.6 this line fills the whole header area. That's a style issue. What else?

**Reply by timunie** [reply]
It's simply breaking MVVM and not supposed to be used that way. If it's your issue, I don't know without a simple sample attached. For me Tabs are working. 

Check DevTools via F12 for VisualTree

**Comment by drei97** [comment]
OK, it breaks MVVM.So what! MVVM is only a concept and not put in stone.

The approach to use code behind is not the root cause for the blue. I also tried with TabItem in the XAML and it is not different to the other way.

Since I have not made any changes in code or styles while going from 11.0.5 to 11.0.6 and vice versa there must be some changes in Avalonia (Theme fluent?) itself. I do not have any other explanation.  

I will prepare a simple example tomorrow.

**Reply by timunie** [reply]
Please don't get me wrong. I'm not saying one must follow MVVM strictly but for TabControl which is an Itemcontrol it can break things due to how virtualization works. I think also having DataTemplates mapped to an ItemsViewModel can improve your code while keeping it flexible (or even more flexible). 

Still not sure where the root cause is. So awaiting your sample to try it on my side. If it's really due to update, I'll open it as bug and add regression tag. That will get a fix soonish. 

Happy coding

**Comment by timunie** [comment]
okay did a simple test App:

ViewModel:
```cs
    public ObservableCollection<TabItem> TabItems { get; } =
    [
        new TabItem(){ Header = 123, Content = 456},
        new TabItem(){ Header = "abc", Content = "def"},
    ];
```

View:
```xml
 <TabControl ItemsSource="{Binding TabItems}"></TabControl>
```

Avalonia-version:
```xml
<PackageReference Include="Avalonia" Version="11.0.6"/>
<PackageReference Include="Avalonia.Desktop" Version="11.0.6"/>
<PackageReference Include="Avalonia.Themes.Fluent" Version="11.0.6"/>
<PackageReference Include="Avalonia.Fonts.Inter" Version="11.0.6"/>
<!--Condition below is needed to remove Avalonia.Diagnostics package from build output in Release configuration.-->
<PackageReference Condition="'$(Configuration)' == 'Debug'" Include="Avalonia.Diagnostics" Version="11.0.6"/>
<PackageReference Include="Avalonia.ReactiveUI" Version="11.0.6"/>
```

result:
![image](https://github.com/AvaloniaUI/Avalonia/assets/47110241/87002cb7-82a5-43a7-8b39-dc5c626120ac)


------

Alternative: 

1 - Add TabItemViewModel
```cs
public class TabItemViewModel : ViewModelBase
{
    public TabItemViewModel(object header, object content)
    {
        Header = header;
        Content = content;
    }

    public object Header { get; set; }
    public object Content { get; set; }
}
```

2 - change collection and implement FuncDataTemplate (check Avalonia.Samples for more info)
```cs
    public ObservableCollection<TabItemViewModel> TabItems { get; } =
    [
        new TabItemViewModel(123, 456),
        new TabItemViewModel("abc", "def"),
    ];

    public FuncDataTemplate<TabItemViewModel> ContentDataTemplate { get; } = new FuncDataTemplate<TabItemViewModel>(
        o => true,
        o => o.Content switch
        {
            int i => new NumericUpDown(){Value = i},
            string s => new TextBox(){Text = s},
            { } x => new TextBlock(){Text = x.ToString()} 
        });
```
 
and finally our View: 
```xml
<TabControl ItemsSource="{Binding TabItems}" 
            ContentTemplate="{Binding ContentDataTemplate}">
    <!-- For the header we can re-use the same template on each item -->
    <TabControl.ItemTemplate>
        <DataTemplate DataType="vm:TabItemViewModel">
            <TextBlock Text="{Binding Header}"></TextBlock>
        </DataTemplate>
    </TabControl.ItemTemplate>
</TabControl>
```

**Reply by drei97** [reply]
I can confirm now, that a freshly set MVVM app with TabItems given in the XAML is working correctly. I tried it again today and the blue issue doesn't appear in 11.0.5 and 11.0.6.

But, for dynamically added TabItems by code behind as in my project, 11.0.5 is fine and 11.0.6 is with the blue TabItem header.

An example is attached. Please note that I am working with Rider under Linux. Maybe there are some differences due to that configuration.

It is interesting to see, that the behavior of dynamically added TabItems has changed with the update. This configuration was working for years now, firstly with WPF and also now with Avalonia. I am not sure if this change was intended.


Your proposal of an alternative must be checked later. I am on vacation next week.

[TabItemTest.zip](https://github.com/AvaloniaUI/Avalonia/files/13683815/TabItemTest.zip)

**Reply by timunie** [reply, marked as answer]
Interesting sample. While I still think this approach is not the best, at lease adding Tabitems from code behind via `MyTabControl.Items.Add` should work, but it does also not work. So I can confirm now it's a bug. Found that the pipe is just too big for some weird reason. Changing it in DevTools results in okay styles 🤷

However, using latest nightly fixes the regression again, so please use nightly or 11.0.5 until next stable release :-)

Have a nice holiday
